### ISSUE TITLE
How to sort a 3D numpy arry

### ISSUE THREAD

**Issue by jinzishuai** [jinzishuai/learn2deeplearn#47]
problem 5 of https://github.com/jinzishuai/learn2deeplearn/blob/master/google_dl_udacity/lesson1/1_notmnist.ipynb

# my code of sorting 4 images
```python
# These are all the modules we'll be using later. Make sure you can import them
# before proceeding further.
from __future__ import print_function
import matplotlib.pyplot as plt
import numpy as np
import data

def look_for_duplicates (dataset):
    print ('looking for duplicates in a dataset of shape: ', dataset.shape)

def sort4images(mytest):
    plt.figure(1)
    plt.subplot(221)
    plt.imshow(mytest[0,:,:])
    plt.subplot(222)
    plt.imshow(mytest[1,:,:])
    plt.subplot(223)
    plt.imshow(mytest[2,:,:])
    plt.subplot(224)
    plt.imshow(mytest[3,:,:])

    mytest.sort(axis=0)
    plt.figure(2)
    plt.subplot(221)
    plt.imshow(mytest[0,:,:])
    plt.subplot(222)
    plt.imshow(mytest[1,:,:])
    plt.subplot(223)
    plt.imshow(mytest[2,:,:])
    plt.subplot(224)
    plt.imshow(mytest[3,:,:])
    plt.show()

letters=np.array(['A', 'B', 'C','D','E','F','G','H','I','J'])
train_dataset, train_labels, valid_dataset, valid_labels, test_dataset, test_labels = data.load_all_data_from_single_pickle_file('../notMNIST.pickle')
# sort the train_set
print('my test results are',letters[train_labels[0:4]])
sort4images(train_dataset[0:4,:,:])
```

## Before Sort
![image](https://user-images.githubusercontent.com/1074685/37882388-2b1a8466-3060-11e8-88d6-797dc80156ef.png)

## After Sort: all messed up
![image](https://user-images.githubusercontent.com/1074685/37882393-31c93442-3060-11e8-8a77-21f11a203814.png)

**Comment by jinzishuai** [comment]
# mytest.sort(axis=-1) the same as mytest.sort(axis=2)
![image](https://user-images.githubusercontent.com/1074685/37882439-951eeeba-3060-11e8-80a5-1913b8d7d914.png)
# mytest.sort(axis=1)
![image](https://user-images.githubusercontent.com/1074685/37882464-c2dabe42-3060-11e8-8463-93a32b68d4dd.png)

**Comment by jinzishuai** [comment]
# Solution: `sorted, indices = np.unique(mytest, axis=0, return_index = True)`

```python
# These are all the modules we'll be using later. Make sure you can import them
# before proceeding further.
from __future__ import print_function
import matplotlib.pyplot as plt
import numpy as np
import data

def look_for_duplicates (dataset):
    print ('looking for duplicates in a dataset of shape: ', dataset.shape)

def sort4images(mytest):
    plt.figure(1)
    plt.subplot(221)
    plt.imshow(mytest[0,:,:])
    plt.subplot(222)
    plt.imshow(mytest[1,:,:])
    plt.subplot(223)
    plt.imshow(mytest[2,:,:])
    plt.subplot(224)
    plt.imshow(mytest[3,:,:])

    #mytest.sort(axis=2)
    sorted, indices = np.unique(mytest, axis=0, return_index = True)
    plt.figure(2)
    plt.subplot(221)
    plt.imshow(sorted[0,:,:])
    plt.subplot(222)
    plt.imshow(sorted[1,:,:])
    plt.subplot(223)
    plt.imshow(sorted[2,:,:])
    plt.subplot(224)
    plt.imshow(sorted[3,:,:])
    plt.show()
    return sorted, indices

letters=np.array(['A', 'B', 'C','D','E','F','G','H','I','J'])
train_dataset, train_labels, valid_dataset, valid_labels, test_dataset, test_labels = data.load_all_data_from_single_pickle_file('../notMNIST.pickle')
# sort the train_set
print('my test results are',letters[train_labels[0:4]])
sorted, indices = sort4images(train_dataset[0:4,:,:])
print('my sorted test results are',letters[train_labels[indices]])
```

**Comment by jinzishuai** [comment]
![image](https://user-images.githubusercontent.com/1074685/37882618-1696398e-3062-11e8-8c89-c532d287a156.png)
